### PR TITLE
[REVIEW] Add support for Dataframe.shift and shift improvements

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR:

- [x] Add support for DataFrame.shift #2390 (https://github.com/rapidsai/cudf/issues/1285#issuecomment-515111994)
- [x] Replaces -1s with null values now.

### CODE REVIEW

**Comment by galipremsagar** [comment]
cc: @beckernick @randerzander @kkraus14

File: python/cudf/cudf/utils/cudautils.py [hunk 1]
```diff
@@ -553,7 +553,7 @@ def gpu_unique_set_insert(vset, sz, val):
 
 
 @cuda.jit
-def gpu_shift(in_col, out_col, N):
+def gpu_shift(in_col, out_col, N, replacement_value):
```
Comment (beckernick) [inline python/cudf/cudf/utils/cudautils.py]: Per our discussion, let's explicitly refer to this as a sentinel value so it's clear that this is not intended to be used generally for replacement.

Comment (kkraus14) [reply python/cudf/cudf/utils/cudautils.py]: Any reason we can't just handle nulls using the utility functions now? We have functions to get and set mask bits.

**Comment by galipremsagar** [comment]
> Should actually handle nulls with bitmask operators in `cudf.utils.utils`

Handling this now. This is Ready for review.

`== 11557 passed, 759 skipped, 85 xfailed, 6 xpassed, 1903 warnings in 378.07 seconds ==`

File: python/cudf/cudf/utils/cudautils.py [hunk 2]
```diff
@@ -553,21 +553,27 @@ def gpu_unique_set_insert(vset, sz, val):
...
             out_col[i] = in_col[i - N]
-        if i >= (in_col.size + N) and i < in_col.size:
-            out_col[i] = -1
+
+    base = i * mask_bitsize
+    for k in range(base, base + mask_bitsize):
+        if k >= in_col.size:
+            break
+        if N > 0 and k >= N:
+            mask_set(mask_bits, k)
```
Comment (jrhemstad) [inline python/cudf/cudf/utils/cudautils.py]: I don't think there's a need to set individual bits here (which invokes a global atomic per bit---expensive).

We're updating contiguous ranges of bits, so instead you can write whole bitmask elements at a time.

Something like this:
```
warp_size = 32
warp_id = i / warp_size
lane_id = i % warp_size
if( warp_id < N / warp_size)
   if(lane_id == 0)
      mask_bits[warp_id] = 0 # zero out the contiguous range of bits
remainder = N % warp_size
if( remainder != 0 and warp_id == (N/warp_size + 1) )
   new_mask = ~(1 << remainder) - 1 # mask with "remainder" LSBs cleared
   mask_bits[warp_id] &= new_mask # bitwise AND with old mask 
```

Comment (kkraus14) [reply python/cudf/cudf/utils/cudautils.py]: @galipremsagar so `warp_size` would be `cuda.warpsize` and `lane_id` would be `cuda.laneid` per here: https://numba.pydata.org/numba-doc/dev/cuda-reference/kernel.html?highlight=warp#numba.cuda.warpsize

**Comment by kkraus14** [comment]
Closing this as @cwharris is implementing in libcudf in #2737

**Pull request closed without merging**